src_feedback/elf.rs

```diff
@@ -0,0 +1,666 @@
+    sh_addr: Elf64_Addr,	/* Section virtual addr at execution */
+    sh_offset: Elf64_Off,	/* Section file offset */
+    sh_size: Elf64_Xword,	/* Size of section in bytes */
+    sh_link: Elf64_Word,	/* Index of another section */
+    sh_info: Elf64_Word,	/* Additional section information */
+    sh_addralign: Elf64_Xword,	/* Section alignment */
+    sh_entsize: Elf64_Xword,	/* Entry size if section holds table */
+}
+
+pub const SHN_UNDEF: u16 = 0;
+
+#[allow(dead_code)]
+pub const STT_NOTYPE: u8 = 0;
+#[allow(dead_code)]
+pub const STT_OBJECT: u8 = 1;
+#[allow(dead_code)]
+pub const STT_FUNC: u8 = 2;
+
+#[derive(Clone)]
+#[repr(C)]
+pub struct Elf64_Sym {
+    st_name: Elf64_Word,	/* Symbol name, index in string tbl */
+    st_info: u8,		/* Type and binding attributes */
+    st_other: u8,		/* No defined meaning, 0 */
+    st_shndx: Elf64_Half,	/* Associated section index */
+    st_value: Elf64_Addr,	/* Value of the symbol */
+    st_size: Elf64_Xword,	/* Associated symbol size */
+}
+
+#[allow(dead_code)]
+impl Elf64_Sym {
+    fn get_type(&self) -> u8 {
+	self.st_info & 0xf
+    }
+
+    fn is_undef(&self) -> bool {
+	self.st_shndx == SHN_UNDEF
+    }
+}
+
+#[allow(dead_code)]
+#[repr(C)]
+struct Elf64_Rel {
+    r_offset: Elf64_Addr,	/* Location at which to apply the action */
+    r_info: Elf64_Xword,	/* index and type of relocation */
+}
+
+#[allow(dead_code)]
+#[repr(C)]
+struct Elf64_Rela {
+    r_offset: Elf64_Addr,	/* Location at which to apply the action */
+    r_info: Elf64_Xword,	/* index and type of relocation */
+    r_addend: Elf64_Sxword,	/* Constant addend used to compute value */
+}
+
+#[allow(dead_code)]
+#[repr(C)]
+union Elf64_Dyn_un{
+    d_val: Elf64_Xword,
+    d_ptr: Elf64_Addr,
+}
+
+#[allow(dead_code)]
+#[repr(C)]
+struct Elf64_Dyn {
+    d_tag: Elf64_Sxword,	/* entry tag value */
+    d_un: Elf64_Dyn_un,
+}
+
+
+fn read_u8(file: &mut File, off: u64, size: usize) -> Result<Vec<u8>, Error> {
+    let mut buf = vec![0; size];
+
+    file.seek(SeekFrom::Start(off))?;
+    file.read_exact(buf.as_mut_slice())?;
+
+    Ok(buf)
+}
+
+fn read_elf_header(file: &mut File) -> Result<Elf64_Ehdr, Error> {
+    const DSZ: usize = mem::size_of::<Elf64_Ehdr>();
+    let mut buf = Box::new([0_u8; DSZ]);
+
+    let buf_m: &mut [u8; DSZ] = buf.borrow_mut();
+    file.read_exact(buf_m)?;
+
+    let ehdr: Box<Elf64_Ehdr> = unsafe {
+	// A complicated type casting hacking!!
+	let ehdr_raw_ptr = (Box::leak(buf) as *mut u8) as *mut Elf64_Ehdr;
+	Box::from_raw(ehdr_raw_ptr)
+    };
+
+    Ok(*ehdr)
+}
+
+fn read_elf_sections(file: &mut File, ehdr: &Elf64_Ehdr) -> Result<Vec<Elf64_Shdr>, Error> {
+    const HDRSIZE: usize = mem::size_of::<Elf64_Shdr>();
+    let off = ehdr.e_shoff as usize;
+    let num = ehdr.e_shnum as usize;
+
+    let mut buf = read_u8(file, off as u64, num * HDRSIZE)?;
+
+    let shdrs: Vec<Elf64_Shdr> = unsafe {
+	let shdrs_ptr = buf.as_mut_ptr() as *mut Elf64_Shdr;
+	buf.leak();
```

Comment:
To confirm my understanding, we need the buf.leak() call here because otherwise, the buf would be freed by Rust, correct?

Reply:
Yes!  The memory would be taken by the Vec created later.
Most of these unsafe codes are about typecasting.